### PR TITLE
DOC: add feature docs for shared access key authentication

### PR DESCRIPTION
**Pull request by stijnmoreels**
Adds features documentation for the shared access key authentication
that can be both configured on global and local level by an attribute
and authentication filter.

 #12

### CODE REVIEW

File: docs/features/shared-access-key.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
+---
+title: "Authentication with shared access keys via ASP.NET Core authentication filters"
+layout: default
+---
+
```
[the remark this answers is not in the document]
Comment (tomkerkhove) [reply docs/features/shared-access-key.md]: For example, what is shared access key authentication - Explains that it looks for an HTTP header and compares the value with your secret.

File: docs/features/shared-access-key.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+This attribute will then add authentication to the route(s) via shared access keys configurable on the attribute itself.
+
+### Usage
+
+The authentication requires a `ICachedSecretProvider` or `ISecretProvider` to be registered in services of the applications (normally in the `Startup` class):
+
+```csharp
+public void ConfigureServices(IServiceCollections services)
+{
+    services.AddScoped<ICachedSecretProvider>(serviceProvider => new MyCachedSecretProvider());
```
[the remark this answers is not in the document]
Comment (stijnmoreels) [reply docs/features/shared-access-key.md]: Ah, yes that would be better!

File: docs/features/shared-access-key.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    services.AddMvc();
+}
+```
+
+After that, you can add the filter to the MVC services:
+
+
+```csharp
```
[the remark this answers is not in the document]
Comment (stijnmoreels) [reply docs/features/shared-access-key.md]: Aha, yes.

File: docs/features/shared-access-key.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (stijnmoreels) [reply docs/features/shared-access-key.md]: Yes.

File: docs/features/shared-access-key.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+This attribute will then add authentication to the route(s) via shared access keys configurable on the attribute itself.
+
+### Usage
+
+The authentication requires a `ICachedSecretProvider` or `ISecretProvider` to be registered in services of the applications (normally in the `Startup` class):
+
+```csharp
+public void ConfigureServices(IServiceCollections services)
+{
+    services.AddScoped<ICachedSecretProvider>(serviceProvider => new CachedSecretProvider(new MySecretProvider()));
```
Comment (tomkerkhove) [inline docs/features/shared-access-key.md]: We can also use our extension but this is ok!

File: docs/index.md [hunk 5]
```diff
@@ -26,6 +26,12 @@ PM > Install-Package Arcus.WebApi.Logging
 
 # Features
 
+## Arcus.WebApi.Security
+
+The `Arcus.WebApi.Logging` package contains functionality that can be incorporated in API projects to easily add security capabilities to an API project.
```
Comment (tomkerkhove) [inline docs/index.md]: Copy paste

File: docs/features/shared-access-key.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
+---
+title: "Authentication with shared access keys via ASP.NET Core authentication filters"
+layout: default
+---
+
+## Arcus.WebApi.Security.SharedAccessKeyAuthentication
```
Comment (tomkerkhove) [inline docs/features/shared-access-key.md]: You can remove this header as we only have this for now.

File: docs/features/shared-access-key.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+The `Arcus.WebApi.Security` package provides an mechanism that uses shared access keys to authenticate users.
+This authentication process consists of two parts:
+
+1. Looks for the configured HTTP request header that contains the shared access key
+2. Matches this value with the configured **Key Vault** secret
```
Comment (tomkerkhove) [inline docs/features/shared-access-key.md]: Actually, this is not true, we will look in whatever secret provider is configured

File: docs/features/shared-access-key.md [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+The package allows two ways to configure this type of authentication mechanmism in an <span>ASP.NET</span> application:
+- [Shared access key attribute](#ArcusWebApiSecuritySharedAccessKeyAuthenticationAttribute) allows local authentication on  `Controller` or method level
+- [Shared access key filter](#ArcusWebApiSecuritySharedAccessKeyAUthenticationFilter) allows global authentication 
+
+## Arcus.WebApi.Security.SharedAccessKeyAuthenticationAttribute
```
Comment (tomkerkhove) [inline docs/features/shared-access-key.md]: Maybe it would make sense to change this to `Enforce shared access key per controller or operation`?

File: docs/features/shared-access-key.md [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    [HttpGet]
+    [Route("authz/shared-access-key")]
+    public Task<IActionResult> AuthorizedGet()
+    {
+        return Task.FromResult<IActionResult>(Ok());
+    }
+}
+```
+
+## Arcus.WebApi.Security.SharedAccessKeyAuthenticationFilter
```
Comment (tomkerkhove) [inline docs/features/shared-access-key.md]: Maybe it would make sense to change this to `Enforce shared access key globally` and move this before our other approach to stimulate this one.

File: docs/features/shared-access-key.md [hunk 10]
```diff
@@ -0,0 +1,69 @@
+---
+title: "Authentication with shared access keys via ASP.NET Core authentication filters"
+layout: default
+---
+
+The `Arcus.WebApi.Security` package provides an mechanism that uses shared access keys to authenticate users.
+This authentication process consists of two parts:
+
+1. Looks for the configured HTTP request header that contains the shared access key
+2. Matches this value with the **Key Vault** secret that is requested via a `ICachedSecretProvider` or `ISecretProvider`
```
Comment (tomkerkhove) [inline docs/features/shared-access-key.md]: It still mentions Key Vault which is not the case, maybe they wrote there own.

I propose the following:
```suggestion
2. Shared access key matches the value with the secret stored, determined via configured secret provider
```

File: docs/index.md (hunk 5 above)
Comment (tomkerkhove) [inline docs/index.md]:
```suggestion
The `Arcus.WebApi.Security` package contains functionality to easily add security capabilities to an API.
```

File: docs/features/shared-access-key.md (hunk 8 above)
Comment (tomkerkhove) [reply docs/features/shared-access-key.md]: Don't forget to update the link above

File: docs/features/shared-access-key.md (hunk 9 above)
Comment (tomkerkhove) [reply docs/features/shared-access-key.md]: Don't forget to update the link above

**Comment by tomkerkhove** [comment]
Looks good to me, but I would like to rename the headers first

File: docs/features/shared-access-key.md (hunk 9 above)
Comment (stijnmoreels) [reply docs/features/shared-access-key.md]: I don't follow, what do you want to update? I changed the short description with `Enforce...`.

Comment (tomkerkhove) [reply docs/features/shared-access-key.md]: Where did you update this?

When I look at the latest version I still see this, but might be doing something wrong then 🤔 

image

Comment (stijnmoreels) [reply docs/features/shared-access-key.md]: Could you rephrase what you want to change from the original?

Comment (tomkerkhove) [reply docs/features/shared-access-key.md]: The screenshot is a good example - The line above my comment (the ##) should be something along my suggestion

**Comment by tomkerkhove** [comment]
Did final tweaks via #41 so we can ship v0.1.0 - Thanks for the good docs @stijnmoreels ! :+1:

**Comment by stijnmoreels** [comment]
> Did final tweaks via #41 so we can ship v0.1.0 - Thanks for the good docs @stijnmoreels ! 👍

Yes, I saw; now I understand :)
